canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -133,42 +137,87 @@ private[apiserver] final class ApiPackageManagementService private (
+      pkgVersion < version
+    }.traverse { case (pId, _) => lookupDar(pId) }
+    .map(_.flatten)
+  }
+
+  private def maximalVersionedDar(dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.get.name
+    val pkgVersion = pkg.metadata.get.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }.maxByOption { case (_, version) =>
+      pkgVersion > version
+    }.traverse { case (pId, _) => lookupDar(pId) }
+    .map(_.flatten)
+  }
+
+  private def typecheckUpgrades(optDar1: Option[(Ref.PackageId, Ast.Package)], optDar2: Option[(Ref.PackageId, Ast.Package)]): Future[Unit] = {
+    optDar1.fold(Future.unit) { case (pkgId1, pkg1) =>
+        Future.fromTry(Typecheck.typecheckUpgrades((pkgId1, pkg1), optDar2))
+    }
+  }
+
+  private def validateUpgrade(upgradingDar: Dar[(Ref.PackageId, Ast.Package)])(implicit
       loggingContext: LoggingContextWithTrace
   ): Future[Unit] = {
-    val upgradingPackageId = upgradingPackage.main._1
+    val (upgradingPackageId, upgradingPackage) = upgradingDar.main
+    val optUpgradingDar = Some((upgradingPackageId, upgradingPackage))
     logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
-    upgradingPackage.main._2.metadata.flatMap(_.upgradedPackageId) match {
+    upgradingPackage.metadata.flatMap(_.upgradedPackageId) match {
       case Some(upgradedPackageId) =>
         for {
-          upgradedArchiveMb <- packagesIndex.getLfArchive(upgradedPackageId)
-          upgradedPackageMb <- Future.fromTry {
-            upgradedArchiveMb.traverse(Decode.decodeArchive(_))
-              .handleError(Validation.handleLfArchiveError)
-          }
-          _ <- {
-            logger.info(s"Package $upgradingPackageId upgrades package id $upgradedPackageId")
-            val upgradeCheckResult = Typecheck.typecheckUpgrades(upgradingPackage.main, upgradedPackageMb)
-            upgradeCheckResult match {
-              case Success(()) => {
-                logger.info(s"Typechecking upgrades for $upgradingPackageId succeeded.")
-                Future(())
-              }
-              case Failure(err: UpgradeError) => {
+          optUpgradedDar <- lookupDar(upgradedPackageId)
+          _ = logger.info(s"Package $upgradingPackageId upgrades package id $upgradedPackageId")
+          _ <- optUpgradingDar.fold {
+            logger.info(s"Typechecking upgrades for $upgradingPackageId failed with following message: CouldNotResolveUpgradedPackageId")
+            Future.failed(Validation.handleUpgradeError(upgradingPackageId, upgradedPackageId, UpgradeError("CouldNotResolveUpgradedPackageId")).asGrpcError)
```

Comment:
We can drop this fold over the optUpgradingDar and just directly access upgradingPackage - saves us this code duplication.